proxy/src/telemetry/metrics/mod.rs

```diff
@@ -597,14 +627,107 @@ impl HyperService for Serve {
                 .with_status(StatusCode::NotFound));
         }
 
-        let body = {
-            let metrics = self.metrics.lock()
-                .expect("metrics lock poisoned");
-            format!("{}", *metrics)
+        let metrics = self.metrics.lock()
+            .expect("metrics lock poisoned")
+            .deref()
+            .clone();
```

Comment:
What is the effect of this, in terms of memory usage? i.e. how large is `metrics`?

Reply:
I'm not sure exactly how large `metrics` would be after aggregating a lot of stats, but could look into profiling it.

My intuition was that the additional allocation from cloning it was worth being able to release the lock earlier., to avoid blocking the aggregate task. If you don't think that's the case, I could rewrite this so that we format the shared `Metrics` rather than cloning, but then the server task will hold the lock for longer.

Reply:
I could be reading wrong, but it looks like this is `Metrics` further up the file, which is some 12 `Metric` values, that each contain an `IndexMap`. If so, this would be cloning a bunch of hash tables.

Reply:
In retrospect, it's probably much better not to clone here, regardless of any advantages of releasing the lock sooner. I'll remove it.